Add Performance Section

dainemawer opened the pull request:
<!--
### Requirements

Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change

<!--
We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts.  Please include screenshots (if appropriate).
-->

This PR adds the Web Vitals section to main navigation and secondary navigation to the Engineering Best Practices site. It also handles some minor CSS tweaks to account for the change in styling. Furthermore, I've added a detail page of content explaining the hows, whats and why's of Core Web Vitals for engineers to follow.

### Benefits

<!-- What benefits will be realized by the code change? -->

New internal guidance for managing and achieving healthy Core Web Vitals across the team.

### Verification Process

<!--
What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., commands you ran, text you typed, buttons you clicked) and describe the results you observed.
-->

- Tested navigation updates on desktop and mobile.

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests passed.

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Applicable Issues

<!-- Enter any applicable Issues here -->
https://github.com/10up/Engineering-Best-Practices/issues/353

### Changelog Entry

<!-- Add sample CHANGELOG.md entry for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related. -->
Added Core Web Vitals content.

Commit:
Added Web Vitals section, updated styles for Web Vitals section, tweaked header CSS to fit Web Vitals link

joesnellpdx commented on _includes/markdown/Web-Vitals.md:
```diff
@@ -0,0 +1,293 @@
+
+> An LCP score of 2.5 seconds or less is considered to be a conducive measurement for good user experience.
+
+LCP is measured in seconds (s) and can be tracked against the follow DOM elements:
```
@dainelmawer 
```suggestion
LCP is measured in seconds (s) and can be tracked against the following DOM elements:
```

dainemawer replied:
Updated, hanks @joesnellpdx !

joesnellpdx commented on web-vitals.md:
```diff
@@ -0,0 +1,23 @@
+---
+page: web-vitals
```
@dainelmawer 

Are we sure we want to be as specific as calling this new page Web Vitals? Should we consider calling it Performance with a current focus on the definition and recommendations pertaining to Web Vitals? Could other performance related resources, tools, best practices come into play here? We could briefly mention performance relating to specific sections in the BP doc are mentioned elsewhere, while general performance best practices regarding perfromance are defined here. 

Something to ponder. I don't want to block the update because of this thought. If this has already been thought about / discussed - disregard my comment and move forward.

dainemawer replied:
@joesnellpdx appreciate your opinion here. Here was my thinking behind it - when I look online across sites that are listing performance improvements specifically around Web Vitals - the URL all kind of centers around the "web-vitals" buzzword. I wanted to make sure that it's absolutely clear that 10up supports Web Vitals in a direct manner. Having a nav item as Web Vitals I think will help with this and we can offer targeted advice around fixing specific web vitals - rather than the general catch-all "performance" term that is already listed under each discipline/language.

I think we can iterate on this approach and clean it up at some point but with the algorithm update approaching fast, I wanted to make sure we were being as transparent as possible.

joesnellpdx replied:
I totally get what you are saying. Per our conversation today, I think the plan is to go with a Performance page and CWV as part of it... but I still like the idea of a direct web vitals url. I'll think about it some more. Maybe after the fact, we could do a post on our external blog and link to our best practices? Just an idea. 

@dainelmawer

dainemawer replied:
Roger that - I think an external blog post would be a great idea.

joesnellpdx commented on _includes/markdown/Web-Vitals.md:
```diff
@@ -0,0 +1,293 @@
+The time it takes the browser to fetch resources like images or videos can also have an effect on LCP:
+* Optimize and compress all images on the site - ensure images are not greater than twice their contained real-estate.
+* Make sure that images are being served over a CDN, you're serving formats like WebP or AVIF and you're using responsive images techniques.
+* For images that find themselves in Hero components, `preload` the image resource ahead of time using `<link rel="preload" href="/path/to/image" as="image" />`. For [responsive images](https://web.dev/preload-responsive-images/) you will need to add the `imagesrcset` and `imagesizes` attribute
```
@dainelmawer 

Aren't they `srcset` and `sizes` attributes?

dainemawer replied:
@joesnellpdx per the spec these two attributes are called `imagesrcset` and `imagesizes` only when placed on a `<link>` tag :)

joesnellpdx replied:
Ahh... I see what you are saying...

Maybe add an example that makes that more clear?:

`  <link rel="preload"
        as="image" 
        imagesrcset="
          image-400.jpg 400w, 
          image-800.jpg 800w, 
          image-1600.jpg 1600w" 
        imagesizes="100vw">`

joesnellpdx commented on _includes/markdown/Web-Vitals.md:
```diff
@@ -0,0 +1,293 @@
+
+Here are some high-level guidelines for ensuring Largest Contentful Paint occurs as fast as possible:
+
+* Are assets (Images, JavaScript, CSS, Video) served over a CDN?
```
@dainelmawer 

for consistency, maybe have all as statements rather some as statements and some as questions? 

Example for this line:
* Serve assets (Images, JavaScript, CSS, Video) over a CDN if viable.

dainemawer replied:
I like it - updated!

joesnellpdx commented on _includes/markdown/Web-Vitals.md:
```diff
@@ -0,0 +1,293 @@
+* All images loaded on the site have a `width` and a `height` attribute. This is because HTML gets parsed before CSS and the browser will reserve space if it knows the dimensions and aspect-ratio of the image.
+* Ensure that ads, iframes and other embeds have a `width` and `height` attribute.
+* As a best practice, do not insert dynamic content into the site without the user performing an action to recieve it, ie "load more" or "click".
+* Ensure that you have a Web Font Loading strategy in place.
```
@dainelmawer 
```suggestion
* Ensure that you have a Web Font Loading strategy in place that mitigates layout shift when fonts are loaded and displayed in the browser (more info on [Web Fonts](#handling-web-fonts) below).
```

dainemawer replied:
Agreed - updated!

joesnellpdx commented on _includes/markdown/Web-Vitals.md:
```diff
@@ -0,0 +1,293 @@
+4. A combination of both 2 and 3 has yielded great improvement to CLS scores.
+5. Consider using `googletag.pubads().collapseEmptyDivs();` to ensure that ad slots that probably wont fill take up no height and width on the page.
+
+#### _Handling Web Fonts_
```
@dainelmawer 
Before you publish this section - I recommend connecting with Ian Del Guidice as he has been doing some font loading experiments relating to performance.

dainemawer replied:
I'll reach out to Ian!

dainemawer replied:
@joesnellpdx I spoke to Ian about this - we were pretty much aligned! No need to add anything more here!

joesnellpdx reviewed (changes requested):
@dainelmawer 

Left come comments and suggestions. 

This is GREAT!

That navigation is pretty squished at that mid breakpoint - but I think it is fine for now (leave as is). 

Let me know if you have any questions / want more feedback.

Commit:
Updates to language and spelling

Commit:
Refactored PR to use Performance and subcategories rather than Core Web Vitals

dainemawer commented:
@brentvr I've updated this PR based on what we spoke about yesterday. @joesnellpdx let me know how this looks to you!

joesnellpdx committed:
Merge branch 'feature-core-web-vitals' of github.com:10up/Engineering-Best-Practices into feature-core-web-vitals

joesnellpdx commented:
@brentvr @dainelmawer 

The new additions from Daine are good at this point (with fonts section may requiring further investigation)

Next steps (from my perspective):

- [ ] Add local nav (bulleted list) at the top of the page covering the different headings / topics (some may link to sections in other areas of the site (i.e. php)
- [ ] Fill in each section with content from Brent's document on Baseline Performance standards
- [ ] Review
- [ ] Define areas of further discovery (i.e. critical css)
- [ ] Approve + deploy
- [ ] Update functional reqs template
- [ ] Optional / possibility: external blog post focusing on Web Vitals and linking back to best practices
- [ ] Begin discussions on next steps (i.e. performance budgets)

joesnellpdx committed:
Increase flyout nav breakpoint to contain all items

dainemawer committed:
Merge pull request #366 from 10up/feature-core-web-vitals--nav-update

Feature core web vitals  nav update

dainemawer commented:
Thanks for the checklist @joesnellpdx - I think this is a good path forward, lets please push to get this out before the end of the month to align with Googles update. @brentvr let me know if you need a hand

Commit:
Add high level baseline performance best practices.

Includes: Philosophy, Caching, Images and Media, Fonts, JavaScript & CSS, Design & UX, Advertising, Systems, Third Party Plugins & Scripts.

brentvr commented:
@dainelmawer @joesnellpdx I've added the baseline performance sections, please can you review? Feedback and edits are of course very welcome. 

A few things to note:
- I've added a Third party embeds and scripts section over a Risks section. 
- There might be very slight duplication between the more high level sections like Images & Media and Fonts with the diagnosis contained in the CWV section. 
- Unrelated but the 8th item in the navigation is showing indented. Can we apply the fix in this PR?

dainemawer commented on _includes/markdown/Performance.md:
```diff
@@ -0,0 +1,374 @@
+*   Images should be served using srcset so that smaller sizes can be shown for smaller viewports.
+*   All images implemented through code should [contain a width and height attribute](https://web.dev/optimize-cls/#images-without-dimensions) this is especially important in avoiding Content Layout Shift issues.
+*   Assets in particular images, should be served through CDN
+*   Images, Videos and iFrames should be lazy loaded. Please note that [WordPress will handle browser level lazy loading](https://make.wordpress.org/core/2021/02/19/lazy-loading-iframes-in-5-7/) using [native lazy load](https://web.dev/native-lazy-loading/). In order for this to take effect the width and height should be set on the tag. 
```
@brentvr we should mention something here about using the `facade` pattern for loading iframes: https://web.dev/third-party-facades/

dainemawer commented on _includes/markdown/Performance.md:
```diff
@@ -0,0 +1,374 @@
+*   Fonts used across the site should be [preloaded](https://web.dev/codelab-preload-web-fonts/). If the fonts are not self hosted then ensure to include the _`crossorigin`_ attribute.
+*   Whenever available, WOFF2 font formats should be used for better compression with a WOFF fallback.
+*   Font files should be minimized (don’t load styles you don’t need)
+*   Fonts should either be served locally or from a single foundry (don’t mix Google fonts and typekit, pick one).
```
Just a spelling correction here: `Typekit`

dainemawer commented on _includes/markdown/Performance.md:
```diff
@@ -0,0 +1,374 @@
+
+<h2 id="fonts" class="anchor-heading">Fonts {% include Util/link_anchor anchor="fonts" %} {% include Util/top %}</h2>
+
+*   Fonts used across the site should be [preloaded](https://web.dev/codelab-preload-web-fonts/). If the fonts are not self hosted then ensure to include the _`crossorigin`_ attribute.
```
Wouldnt `crossorigin` be more of a security thing than a performance thing?

dainemawer commented on _includes/markdown/Performance.md:
```diff
@@ -0,0 +1,374 @@
+
+*   Fonts used across the site should be [preloaded](https://web.dev/codelab-preload-web-fonts/). If the fonts are not self hosted then ensure to include the _`crossorigin`_ attribute.
+*   Whenever available, WOFF2 font formats should be used for better compression with a WOFF fallback.
+*   Font files should be minimized (don’t load styles you don’t need)
```
We should probably clarify here that Font CSS should be minimized / minified - the actual font files would need to be subset which is a whole other story.

dainemawer commented on _includes/markdown/Performance.md:
```diff
@@ -0,0 +1,374 @@
+
+<h2 id="javascript-and-css" class="anchor-heading">JavaScript and CSS {% include Util/link_anchor anchor="javascript-and-css" %} {% include Util/top %}</h2>
+
+*   Use “Mobile First” build concepts so the browser doesn’t have to render extra CSS at load time.
```
This is very arbitrary just in the fact that we mention `Mobile First` and `render extra CSS` - we could probably mention something about ensuring critical css or rendering blocking css.

dainemawer commented on _includes/markdown/Performance.md:
```diff
@@ -0,0 +1,374 @@
+*   Use “Mobile First” build concepts so the browser doesn’t have to render extra CSS at load time.
+*   All JavaScript and CSS should be minified.
+*   Standalone site features should be broken off into isolated entry points so we don’t have to load more CSS/JS on pages that will never use it. 
+*   No unnecessary libraries should be loaded. If something is not being used, do not load it, if only a small part of a library is needed, consider custom building or loading only the part that's needed.
```
We should tighten this up - its not always as clear as just forgoing unecessary libraries. Some 3rd-party scripts will embed libraries beyond our control and we cant do much about that.

dainemawer committed:
Update navigation items and positioning as well as content amendments

dainemawer commented:
@joesnellpdx @brentvr I've made updates to the content here as well as:

- Fixed submenu item alignment
- Fixed main navigation positioning and alignment.

Commit:
Fix typos in the Performance page.

joesnellpdx committed:
A few syntactical updates to the text.

tylercherpak commented:
I think this section looks really great! Thanks @dainelmawer and everyone for your hard work. This is an excellent addition to the best practices. 😄

tlovett1 commented on _includes/markdown/Performance.md:
```diff
@@ -0,0 +1,371 @@
+
+There are multiple factors to take into account when considering performance on a project. Third party embeds or scripts can often have a negative impact on performance, but in some cases the inherent value that they bring may justify the performance impact incurred. 
+
+Performance should be a consideration throughout the entire engineering process. Significant performance gains can be achieved through attention to detail and following the best practices documented below.
```
The key thing we want to document is our baseline performance standards - things we do on all projects. To me this just said these are some best practices but not explicitly these are the baseline performance standards for all our projects.

brentvr replied:
@tlovett1 updated, let me know if you feel this needs more rewording.

tlovett1 commented on _includes/markdown/Performance.md:
```diff
@@ -0,0 +1,371 @@
+*   Fonts used across the site should be [preloaded](https://web.dev/codelab-preload-web-fonts/).
+*   Whenever available, WOFF2 font formats should be used for better compression with a WOFF fallback.
+*   Subset Font files if you have them available locally.
+*   Investigate using `unicode-range` to subset fonts if they are being served locally or through Google Fonts.
```
Maybe include a link to a related article here?

dainemawer replied:
Updated, thanks @tlovett1

tlovett1 commented on _includes/markdown/Performance.md:
```diff
@@ -0,0 +1,371 @@
+<h2 id="third-party-plugins-and-scripts" class="anchor-heading">Third Party Plugins and Scripts {% include Util/link_anchor anchor="third-party-plugins-and-scripts" %} {% include Util/top %}</h2>
+
+*   Third party plugins can play a part in poor site performance. Always audit a plugin for performance issues before adding it to a project. Paying careful attention to the server side impacts of slow non-cached queries and how data is being stored to the weight of the JavaScript and CSS being included on every page.
+*   Third party scripts, particularly those being loaded for analytics, user interactivity and advertising can have a major impact on site performance. Oftentimes, these scripts may be loaded with Google Tag Manager. Ensure that when engineering a solution, a discussion is being had around what scripts will be used and test the site performance with all scripts loaded.
```
Can we reword? I'm having trouble understanding the first sentence.

brentvr replied:
@tlovett1 updated, thanks

dainemawer committed:
Added link to uni-code range point

Commit:
Updated Performance section methodology statement and rephrased third party scripts sub section.

tlovett1 committed:
Language tweaks

tlovett1 committed:
Heading tags
